File: 41-FirstMissingPositive/41-FirstMissingPositive.java
// Last updated: 8/13/2025, 8:23:58 PM
class Solution {
    public int firstMissingPositive(int[] nums) {
        int n = nums.length;
        int i=0;
        while(i<n){
            if(nums[i] == i+1){
                i++;
                continue;
            }

            if(nums[i] <= 0 || nums[i] > n){
                i++;
                continue;
            }

            else{
                int idx = nums[i]-1;
                if(nums[idx] == nums[i]){
                    i++;
                }
                else{
                    int temp = nums[idx];
                    nums[idx] = nums[i];
                    nums[i] = temp;
                }
            }
        }

        for(int j=0;j<n;j++){
            if(j!=nums[j]-1){
                return j+1;
            }
        }
        return n+1;
    }
}

File: 41-FirstMissingPositive/41-FirstMissingPositive.py
# Last updated: 8/13/2025, 8:23:51 PM
class Solution:
    def firstMissingPositive(self, nums: List[int]) -> int:
        n = len(nums)

        i = 0 
        while ( i < n) : 
            if nums[i] <1 or nums[i] >n-1 or nums[i]-1 == i: 
                i+=1
            
            else : 
                correctIndex = nums[i]-1
                if nums[correctIndex] == nums[i] : i+=1
                else : [nums[i] , nums[correctIndex]] = [nums[correctIndex] , nums[i]]
        print(nums)
        for i in range(n) : 
            if nums[i]-1 != i : 
                return i+1
        return n+1
